src/Core/Checkout/Cart/CartRuleLoader.php

```diff
@@ -327,4 +335,32 @@ private function validateTaxFree(SalesChannelContext $context, Cart $cart, CartB
 
         return $cart;
     }
+
+    private function translateCartErrors(ErrorCollection $errorCollection, SalesChannelContext $context): void
+    {
+        foreach ($errorCollection as $error) {
+            $parameters = [];
+
+            foreach ($error->getParameters() as $key => $value) {
+                $parameters['%' . $key . '%'] = $value;
+            }
+
+            if ($error->getRoute() instanceof ErrorRoute) {
+                $parameters['%url%'] = $this->router->generate(
+                    $error->getRoute()->getKey(),
+                    $error->getRoute()->getParams(),
+                    UrlGeneratorInterface::ABSOLUTE_PATH
+                );
+            }
```

Comment:
this needs still to be happend in storefront ONLY. the headless clients doesn't know storefront bundle routes. It's better when they have the key + params only and redirect based on that to their client page